Microsoft.HealthVault.Fhir/Transformers/AllergyIntoleranceToHealthVault.cs

```diff
@@ -0,0 +1,155 @@
+//
+// The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED *AS IS*, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+using System.Linq;
+using Hl7.Fhir.Model;
+using Hl7.Fhir.Support;
+using Microsoft.HealthVault.Fhir.Constants;
+using Microsoft.HealthVault.ItemTypes;
+using static Hl7.Fhir.Model.AllergyIntolerance;
+
+namespace Microsoft.HealthVault.Fhir.Transformers
+{
+    public static class AllergyIntoleranceToHealthVault
+    {
+        public static Allergy ToHealthVault(this AllergyIntolerance allergyIntolerance)
+        {
+            if (allergyIntolerance.Code.IsNullOrEmpty())
+            {
+                throw new System.ArgumentException($"Can not transform a {typeof(AllergyIntolerance)} with no code into {typeof(Allergy)}");
+            }
+
+            var allergy = allergyIntolerance.ToThingBase<Allergy>();
+
+            string allergenType = string.Empty;
+            var allergyExtension = allergyIntolerance.GetExtension(HealthVaultExtensions.Allergy);
+            if (allergyExtension != null)
+            {
+                allergenType = allergyExtension.GetStringExtension(HealthVaultExtensions.AllergenType);
+                allergy.AllergenCode = allergyExtension.GetExtensionValue<CodeableConcept>(HealthVaultExtensions.AllergenCode)?.ToCodableValue();
+                allergy.Treatment = allergyExtension.GetExtensionValue<CodeableConcept>(HealthVaultExtensions.AllergyTreatement)?.ToCodableValue();
+            }
+
+            var coding = allergyIntolerance.Code.Coding.FirstOrDefault();
+
+            if (HealthVaultVocabularies.SystemContainsHealthVaultUrl(coding.System))
+            {
+                allergy.Name = allergyIntolerance.Code.ToCodableValue();
+            }
+            else
+            {
+                allergy.SetAllergyName(
+                    coding.Display,
+                    coding.Code,
+                    HealthVaultVocabularies.Fhir,
+                    coding.System,
+                    coding.Version);
+            }
+
+            if (allergyIntolerance.Reaction != null && allergyIntolerance.Reaction.Count > 0)
+            {
+                var code = allergyIntolerance.Reaction.FirstOrDefault().Manifestation.FirstOrDefault().Coding.First();
```

Comment:
I believe coding has a cardinality of 0-* usually, so we probably want to do `FirstOrDefault()` and null check it before using.